plugins/filter_aws/aws.c

```diff
@@ -572,13 +573,6 @@ static int get_ec2_tags(struct flb_filter_aws *ctx)
     ret = get_ec2_tag_keys(ctx);
     if (ret < 0) {
         flb_filter_aws_tags_destroy(ctx);
-        if (ret == -2) {
-            /* -2 means there are no tags, */
-            /* to avoid requesting ec2 tags repeatedly for each flush */
-            /* it marks fetching tags as done */
-            ctx->tags_fetched = FLB_TRUE;
-            return 0;
-        }
         return ret;
```

Comment:
I think we want to keep this bit right? If we get an explicit 404 then tags aren't enabled so we should just log a useful error message and move on- we should not retry because we expect we will keep getting a 404? Or is your thinking that the user might go and enable tags on the instance so we should keep retrying so that FLB will work without restarting once they fix the IMDS settings?

Reply:
We want to keep the same logic for handling 404 errors. However, because the new logic in the proposal for ALL errors is to not retry, therefore we no longer need this if and be specific about 'not found' errors.

Fluent Bit would still produce a helpful message about the root cause of the errors.
```
[2023/06/13 01:33:37] [ warn] [filter:aws:aws.0] EC2 instance metadata tag request returned 404. This likely indicates your instance has no tags or the EC2 tagging metadata API is not enabled
[2023/06/13 01:33:37] [error] [filter:aws:aws.0] Failed to get instance EC2 Tags, will not retry
```

^^ this is based on the the example below

```
mwarzynski@f594bbb3cae9:/fluent-bit-dev/build-dev$ ./bin/fluent-bit -c ./fluent-bit.conf
Fluent Bit v2.1.5
* Copyright (C) 2015-2022 The Fluent Bit Authors
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2023/06/13 01:33:37] [ info] Configuration:
[2023/06/13 01:33:37] [ info]  flush time     | 1.000000 seconds
[2023/06/13 01:33:37] [ info]  grace          | 5 seconds
[2023/06/13 01:33:37] [ info]  daemon         | 0
[2023/06/13 01:33:37] [ info] ___________
[2023/06/13 01:33:37] [ info]  inputs:
[2023/06/13 01:33:37] [ info]      stdin
[2023/06/13 01:33:37] [ info] ___________
[2023/06/13 01:33:37] [ info]  filters:
[2023/06/13 01:33:37] [ info]      aws.0
[2023/06/13 01:33:37] [ info] ___________
[2023/06/13 01:33:37] [ info]  outputs:
[2023/06/13 01:33:37] [ info]      stdout.0
[2023/06/13 01:33:37] [ info] ___________
[2023/06/13 01:33:37] [ info]  collectors:
[2023/06/13 01:33:37] [ info] [fluent bit] version=2.1.5, commit=690a3009f6, pid=12879
[2023/06/13 01:33:37] [debug] [engine] coroutine stack size: 24576 bytes (24.0K)
[2023/06/13 01:33:37] [ info] [storage] ver=1.4.0, type=memory, sync=normal, checksum=off, max_chunks_up=128
[2023/06/13 01:33:37] [ info] [cmetrics] version=0.6.1
[2023/06/13 01:33:37] [ info] [ctraces ] version=0.3.1
[2023/06/13 01:33:37] [ info] [input:stdin:stdin.0] initializing
[2023/06/13 01:33:37] [ info] [input:stdin:stdin.0] storage_strategy='memory' (memory only)
[2023/06/13 01:33:37] [debug] [stdin:stdin.0] created event channels: read=21 write=22
[2023/06/13 01:33:37] [debug] [input:stdin:stdin.0] buf_size=16000
[2023/06/13 01:33:37] [debug] [imds] using IMDSv1
[2023/06/13 01:33:37] [debug] [http_client] not using http_proxy for header
[2023/06/13 01:33:37] [debug] [imds] using IMDSv1
[2023/06/13 01:33:37] [debug] [http_client] not using http_proxy for header
[2023/06/13 01:33:37] [debug] [imds] using IMDSv1
[2023/06/13 01:33:37] [debug] [http_client] not using http_proxy for header
[2023/06/13 01:33:37] [debug] [aws_client] (null): http_do=0, HTTP Status: 404
[2023/06/13 01:33:37] [ warn] [filter:aws:aws.0] EC2 instance metadata tag request returned 404. This likely indicates your instance has no tags or the EC2 tagging metadata API is not enabled
[2023/06/13 01:33:37] [error] [filter:aws:aws.0] Failed to get instance EC2 Tags, will not retry
[2023/06/13 01:33:37] [debug] [stdout:stdout.0] created event channels: read=24 write=25
[2023/06/13 01:33:37] [ info] [sp] stream processor started
[2023/06/13 01:33:37] [ info] [output:stdout:stdout.0] worker #0 started
{"test":1}
[2023/06/13 01:33:40] [debug] [input chunk] update output instances with new chunk size diff=86, records=1, input=stdin.0
[2023/06/13 01:33:40] [debug] [task] created task=0x7f6f58024970 id=0 OK
[2023/06/13 01:33:40] [debug] [output:stdout:stdout.0] task_id=0 assigned to thread #0
[0] stdin.0: [[1686620020.081632328, {}], {"test"=>1, "az"=>"us-east-1a", "ec2_instance_id"=>"i-0e66fc7f9809d7168"}]
[2023/06/13 01:33:40] [debug] [out flush] cb_destroy coro_id=0
[2023/06/13 01:33:40] [debug] [task] destroy task=0x7f6f58024970 (task_id=0)
{"test":2}
[2023/06/13 01:33:43] [debug] [input chunk] update output instances with new chunk size diff=86, records=1, input=stdin.0
[2023/06/13 01:33:43] [debug] [task] created task=0x7f6f58028c60 id=0 OK
[2023/06/13 01:33:43] [debug] [output:stdout:stdout.0] task_id=0 assigned to thread #0
[0] stdin.0: [[1686620023.171430050, {}], {"test"=>2, "az"=>"us-east-1a", "ec2_instance_id"=>"i-0e66fc7f9809d7168"}]
[2023/06/13 01:33:43] [debug] [out flush] cb_destroy coro_id=1
[2023/06/13 01:33:43] [debug] [task] destroy task=0x7f6f58028c60 (task_id=0)
{"test":3}
[2023/06/13 01:33:46] [debug] [input chunk] update output instances with new chunk size diff=86, records=1, input=stdin.0
[2023/06/13 01:33:46] [debug] [task] created task=0x7f6f58028ea0 id=0 OK
[2023/06/13 01:33:46] [debug] [output:stdout:stdout.0] task_id=0 assigned to thread #0
[0] stdin.0: [[1686620026.168247849, {}], {"test"=>3, "az"=>"us-east-1a", "ec2_instance_id"=>"i-0e66fc7f9809d7168"}]
[2023/06/13 01:33:46] [debug] [out flush] cb_destroy coro_id=2
[2023/06/13 01:33:46] [debug] [task] destroy task=0x7f6f58028ea0 (task_id=0)
^C[2023/06/13 01:33:47] [engine] caught signal (SIGINT)
[2023/06/13 01:33:47] [ warn] [engine] service will shutdown in max 5 seconds
[2023/06/13 01:33:47] [ info] [engine] service has stopped (0 pending tasks)
[2023/06/13 01:33:47] [ info] [output:stdout:stdout.0] thread worker #0 stopping...
[2023/06/13 01:33:47] [ info] [output:stdout:stdout.0] thread worker #0 stopped
```

and it only did one request for instance tags to the metadata server:
```
mwarzynski@f594bbb3cae9:/fluent-bit-dev/build-dev$ python3 ec2_server.py
127.0.0.1 - - [13/Jun/2023 01:33:37] "GET /latest/meta-data/instance-id/ HTTP/1.1" 200 -
127.0.0.1 - - [13/Jun/2023 01:33:37] "GET /latest/meta-data/placement/availability-zone/ HTTP/1.1" 200 -
127.0.0.1 - - [13/Jun/2023 01:33:37] "GET /latest/meta-data/tags/instance HTTP/1.1" 404 -
```

Reply:
Cool, makes sense!